### ISSUE TITLE
Agents stop working after 20 minutes

### ISSUE THREAD

**Issue by laurentopia** [GameDevEducation/UnityTutorial_SimsStyleAI#1]
![image](https://user-images.githubusercontent.com/26075468/217131372-6674999e-0264-401b-b7c3-a819a97d3d6d.png)
tracing
![image](https://user-images.githubusercontent.com/26075468/217131531-b81ed5ce-1078-43f1-8192-6718108b6d21.png)
![image](https://user-images.githubusercontent.com/26075468/217131721-a2686683-3384-4ceb-b2f5-6372e5b754ef.png)
perform not called
seems that in certain conditions the agent unlocks smart object but doesn't release current interaction, the odd aspect of this is that it takes a while before breaking and since OnInteractionFinished is passed as callback and that never breaks in my experience it is weird™

**Comment by laurentopia** [comment]
and in the case of the other agent
![image](https://user-images.githubusercontent.com/26075468/217137430-d057ee72-29d1-46b0-8df8-aae2e64b777e.png)
so nothing gets done, it seems, but that's not it, i think this value gets set to performer when it's reached destination so this could be (yet another) unity navmesh related bug.

**Comment by laurentopia** [comment]
And when adding more agents the problems takes less time to show itself.

**Comment by IainMcManus** [comment]
Cheers for the detailed rundown, I'll try and look into it this weekend.

**Comment by IainMcManus** [comment]
@laurentopia sorry for how long this took. I've submitted what I think should fix the issue. With this change I've had 5 AIs run for over an hour without stalling. I think there's an issue with the Unity Navmesh agent where after a time of trying to avoid another AI it stalls out and clears the path info. I'll keep the issue open for now as I'd like to dig more into what's happening on the avoidance side and see if I can implement a fix for the underlying issue.

**Comment by laurentopia** [comment]
Fantastic, thanks for spending time investigating, I'll look at the change. Even with other pathers you run into stall, sometimes. It might be a good failsafe to add a timeout that calls FindNewBestTask() when stall is detected?

**Comment by laurentopia** [comment]
I recommend this instead of .y=0, decoupling the AI from game geometry constraints.

![image](https://user-images.githubusercontent.com/26075468/221045990-460ba4b7-2242-47c2-9a6f-b7c0c201bdc4.png)

**Comment by IainMcManus** [comment]
I switched it over to a similar approach that checks the height delta and the horizontal.

else if (LinkedAgent.hasPath == false)
{
    Vector3 vecToDestination = Destination - transform.position;
    float heightDelta = Mathf.Abs(vecToDestination.y);
    vecToDestination.y = 0f;

    atDestination = heightDelta < LinkedAgent.height && vecToDestination.magnitude <= DestinationReachedThreshold;
}

**Comment by laurentopia** [comment]
That'll only work if the plane of action is perpendicular to y.

**Comment by IainMcManus** [comment]
Yep, that's intentional. I do want to eventually support handling any surface orientation for the AIs but I want to do that in a single commit for clarity.

**Comment by laurentopia** [comment]
That makes sense, I'll keep an eye out for when you do that.
